### PR TITLE
pkg/p2p: add some helper structs for integrating p2p

### PR DESCRIPTION
**Pull request by liuzix**
- Added some adapter logic to make `pkg/p2p` in `tiflow` easier to be used and mocked in Dataflow Engine.

### CODE REVIEW

File: pkg/p2p/message_handler_manager.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+	CheckError(ctx context.Context) error
+
+	// Clean unregisters all existing handlers.
+	Clean(ctx context.Context) error
+
+	// SetTimeout sets the timeout for handler operations.
+	// A timeout is needed because the underlying handler operations are
+	// asynchronous in the MessageServer.
+	SetTimeout(timeout time.Duration)
```
Comment (hanfei1991) [inline pkg/p2p/message_handler_manager.go:33]: If we expose this function, does it mean we have demand of setting timeout on the fly?

File: pkg/p2p/message_handler_manager.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+func (m *messageHandlerManagerImpl) RegisterHandler(
+	ctx context.Context,
+	topic Topic,
+	tpi TypeInformation,
+	fn HandlerFunc,
+) (bool, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.topics[topic]; ok {
```
Comment (hanfei1991) [inline pkg/p2p/message_handler_manager.go:71]: Will it possible to replace the topic, or do we have to remove it at first before we replace it?

File: pkg/p2p/server.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+	wg.Go(func() (err error) {
+		defer log.L().ErrorFilterContextCanceled("message server exited", zap.Error(err))
+		return errors.Trace(s.messageServer.Run(ctx))
+	})
+
+	wg.Go(func() (err error) {
+		defer func() {
+			// TODO (zixiong) filter out expected harmless errors.
+			log.L().Debug("grpc server exited", zap.Error(err))
```
Comment (hanfei1991) [inline pkg/p2p/server.go:95]: Will it always be nil if you don't assign it explictly?

File: pkg/p2p/message_handler_manager.go (hunk 1 above)
Comment (liuzix) [reply pkg/p2p/message_handler_manager.go:33]: - Allowing setting the timeout in the constructor will make the constructor function too complicated.
- Making the timeout a constant will make unit testing difficult.
Although we do not have demand for resetting the timeout on the fly, I think exposing a method is the better way.

File: pkg/p2p/message_handler_manager.go (hunk 2 above)
Comment (liuzix) [reply pkg/p2p/message_handler_manager.go:71]: We have to remove it as far as `MessageServer` is concerned. The implementation of the `MessageServer` does not support replacing it without removing it first. Can you think of a situation where exposing a `Replace` method is helpful?

File: pkg/p2p/server.go (hunk 3 above)
Comment (liuzix) [reply pkg/p2p/server.go:95]: No. The deferred function can see the return value.

File: pkg/p2p/message_handler_manager.go (hunk 2 above)
Comment (hanfei1991) [reply pkg/p2p/message_handler_manager.go:71]: No, I can't figure out this senario either. That's fine

File: pkg/p2p/message_handler_manager.go (hunk 1 above)
Comment (hanfei1991) [reply pkg/p2p/message_handler_manager.go:33]: ok let it be

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**